api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+					"_id":   nil,
+					"total": bson.M{"$sum": "$amount_nano_fil"},
+					"avg":   bson.M{"$avg": "$amount_nano_fil"},
+					"max":   bson.M{"$max": "$amount_nano_fil"},
+					"min":   bson.M{"$min": "$amount_nano_fil"},
+				}},
+			},
+		}},
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calling aggregate: %v", err)
+	}
+	var res []report
+	if err = cursor.All(ctx, &res); err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding cursor results: %v", err)
+	}
+	if len(res) != 1 {
+		return nil, status.Errorf(codes.Internal, "unexpected number of aggregate results: %v", len(res))
+	}
+
+	r := res[0]
+
+	resp := &pb.SummaryResponse{}
+
+	if len(r.All) == 1 {
+		resp.CountTxns = r.All[0].Count
+	}
+
+	for _, state := range r.ByMessageState {
+		switch pb.MessageState(state.ID.(int32)) {
+		case pb.MessageState_MESSAGE_STATE_PENDING:
+			resp.CountPending = state.Count
+		case pb.MessageState_MESSAGE_STATE_ACTIVE:
+			resp.CountActive = state.Count
+		case pb.MessageState_MESSAGE_STATE_FAILED:
+			resp.CountFailed = state.Count
+		}
+	}
+
+	if len(r.Waiting) == 1 {
+		resp.CountWaiting = r.Waiting[0].Count
+	}
+
+	resp.CountFromAddrs = int64(len(r.UniqueFrom))
+	resp.CountToAddrs = int64(len(r.UniqueTo))
+
+	if len(r.SentFilStats) == 1 {
+		resp.TotalNanoFilSent = r.SentFilStats[0].Total
+		resp.AvgNanoFilSent = r.SentFilStats[0].Avg
+		resp.MaxNanoFilSent = r.SentFilStats[0].Max
+		resp.MinNanoFilSent = r.SentFilStats[0].Min
+	}
+
+	return resp, nil
+}
+
+type waitResult struct {
+	txn txn
+	err error
+}
+
+func (s *Service) wait(tx txn) chan waitResult {
+	s.waitingLck.Lock()
+	defer s.waitingLck.Unlock()
+
+	waitCh, found := s.waiting[tx.ID]
+	if found {
+		return waitCh
+	}
+
+	ch := make(chan waitResult)
+	s.waiting[tx.ID] = ch
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), s.config.MessageWaitTimeout)
+
+		client, closeClient, err := s.clientBuilder(ctx)
+		defer func() {
+			closeClient()
+			close(ch)
+			cancel()
+		}()
+		if err != nil {
+			ch <- waitResult{err: fmt.Errorf("creating lotus client: %v", err)}
+			return
+		}
+
+		lastestMsgCid, err := tx.latestMsgCid()
+		if err != nil {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("getting latest message cid from txn: %v", err)
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			return
+		}
+		c, err := cid.Decode(lastestMsgCid.Cid)
+		if err != nil {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("decoding latest message cid for txn: %v", err)
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			return
+		}
```

Comment:
If for some strange reason we can't make sense of the cid associated with the txn, that is a fatal/final error condition.